### PR TITLE
fix: don't buffer when extracting binary

### PR DESCRIPTION
**Pull request by sobolk**
#### Description of changes

Pipes the file instead of buffering in memory at binary installation.

#### Description of how you validated changes

E2E tests. I.e. `amplify_sudo_install_test`

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli-npm/binary.ts [hunk 1]
```diff
@@ -153,27 +153,19 @@ export class Binary {
...
-      if (chunks.length) {
-        const data = Buffer.concat(chunks);
-        fs.writeFileSync(this.binaryPath, data, {
-          mode: 0o755,
-        });
-      }
-    });
     return extract;
   }
 }
```
Comment (sobolk) [inline packages/amplify-cli-npm/binary.ts:171]: there's strong assumption in this code that TAR has only one file.

**Review by edwardfoyle — DISMISSED** [review]

File: .circleci/config.base.yml [hunk 2]
```diff
@@ -454,6 +454,10 @@ jobs:
             startLocalRegistry "$(pwd)/.circleci/verdaccio.yaml"
             setSudoNpmRegistryUrlToLocal
             changeSudoNpmGlobalPath
+            # limit memory for new processes to 1GB
+            # this is to make sure that install can work on small VMs
+            # i.e. not buffer content in memory while installing binary
+            ulimit -Sv 1000000
```
Comment (sobolk) [inline .circleci/config.base.yml:460]: **12.0.0 release**
image

**After applying fix**
image

File: packages/amplify-cli-npm/binary.ts [hunk 3]
```diff
@@ -121,7 +121,10 @@ export class Binary {
     console.log(`Downloading release from ${getCompressedBinaryUrl()}`);
     try {
       const res = await axios({ url: getCompressedBinaryUrl(), responseType: 'stream' });
-      await pipeline(res.data, createGunzip(), this.extract());
+      // Dummy array to collect a promise from nested pipeline that extracts tar content to a file.
+      const extractPromiseCollector: Array<Promise<void>> = [];
+      await pipeline(res.data, createGunzip(), this.extract(extractPromiseCollector));
+      await extractPromiseCollector[0];
```
Comment (edwardfoyle) [inline packages/amplify-cli-npm/binary.ts:127]: why only await the first promise and not Promise.allSettled?

Comment (sobolk) [reply packages/amplify-cli-npm/binary.ts:127]: if we get here there's only one promise, but `Promise.all` would do too.

Comment (sobolk) [reply packages/amplify-cli-npm/binary.ts]: Changed. I used `Promise.all` not `Promise.allSettled` . This should throw if extracting fails.

File: .circleci/config.base.yml (hunk 2 above)
Comment (awsluja) [reply .circleci/config.base.yml:480]: nice

**Review by awsluja — APPROVED** [review]
looks good, only nit would be to add a note why we had to go this route (spawnSync was executing prematurely because tar.extract callbacks were not being awaited; this approach collects those nested callbacks and forces the process to await all) or something like that

**Review by edwardfoyle — DISMISSED** [review]

**Review by awsluja — APPROVED** [review]

**Review by edwardfoyle — APPROVED** [review]

**Pull request merged**